modules/ROOT/pages/reliability-patterns.adoc

```diff
@@ -0,0 +1,162 @@
+
+If you have a web application that uses a non-transactional connector, such as HTTP, you must follow a reliability pattern to ensure reliable messaging for your application.
+
+[NOTE]
+Note that VM file persistency is disabled on clusters, so VM endpoints persist in-memory in a clustered topology.
+
+== Reliability Patterns
+
+A reliability pattern is a design that results in reliable messaging for an application, even if the application receives messages from a non-transactional transport. A reliability pattern couples a reliable acquisition flow with an application logic flow, as shown in the following diagram:
+
+image::reliability-pattern.png[Reliability_Pattern]
+
+The reliable acquisition flow (the left-hand part of the diagram) delivers a message reliably from a Message Source that does not implement transactions to an outbound operation of a connector that implements transactions. The operation can be of any type of the transactional endpoints, such as VM or JMS. If the reliable acquisition flow cannot deliver the message, it ensures that the message isn't lost:
+
+* For socket-based connections like HTTP, this means returning an "unsuccessful request" response to the client so that the client can retry the request.
+* For resource-based connections like File or FTP, it means not deleting the file, so that it can be reprocessed.
+
+The application logic flow (the right-hand side of the diagram) delivers the message from the Message Source that uses a transactional connector to the business logic for the application.
+
+== Comparing Endpoints in Reliability Patterns
+
+You can couple the reliable acquisition flow with any type of transactional endpoint. You don't have to use a VM endpoint in a reliability pattern. The following diagram illustrates a reliability pattern where the reliable acquisition flow delivers a message to a JMS endpoint:
+
+image::reliabilitypatternwithjms.png[ReliabilityPatternwithJMS]
+
+The VM connector uses in-memory queues by default, this makes it much faster than other transactional endpoints such as JMS.
+
+The following table lists the advantages and disadvantages of using VM, JMS, and JDBC endpoints in a reliable acquisition flow:
+
+[%header,cols="10a,60a,30a"]
+|===
+|Endpoint |Advantage |Disadvantage
+|VM (standalone Mule)
+|
+* Faster than JMS because everything is in memory and in-process.
+* Provided ready to use with Mule. There is no need to purchase additional software.
+|
+* Data is lost when Mule exits.
+* Non-clustered.
+|File-based VM (standalone Mule)
+|
+* Provided ready to use with Mule. There is no need to purchase additional software.
+|
+* Non-clustered.
+|VM (Mule HA cluster)
+|
+* Faster than JMS because everything is in shared memory.
+* Provided ready to use with Mule. There is no need to purchase additional software.
+|
+* Data is lost when the entire Mule cluster exits.
+|JDBC
+|
+* Depending on resources available, can hold a very large amount of information.
```

Comment:
It seems like most of the sentences in the table are complete sentences (not fragments) with periods at the end. For line 66, add JDBC so it's a complete sentence, ie, ... JDBC can hold a large amount of information." (I would delete "very" for minimalism)